### PR TITLE
[spirv] Add LLVM-style RTTI to instruction classes

### PR DESCRIPTION
**Pull request by antiagainst**
Also reordered and polished instruction classes

### CODE REVIEW

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h [hunk 1]
```diff
@@ -14,277 +14,126 @@
...
+    IK_Variable,        // OpVariable
+
+    // Function structure kinds
+
+    IK_FunctionParameter, // OpFunctionParameter
+
+    IK_LoopMerge,      // OpLoopMerge
+    IK_SelectionMerge, // OpSelectionMerge
+
+    IK_Branch,            // OpBranch
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/SpirvInstruction.h]: I think it is worth to add a comment here that the order of these enums matters (should change carefully)
e.g. our code does >=IK_Branch && <= IK_Unreachable.

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h [hunk 2]
```diff
@@ -14,277 +14,126 @@
...
+    IK_Atomic,                    // OpAtomic*
+    IK_Barrier,                   // Op*Barrier
+    IK_BinaryOp,                  // Binary operations
+    IK_BitFieldExtract,           // OpBitFieldExtract
+    IK_BitFieldInsert,            // OpBitFieldInsert
+    IK_Composite,                 // Op*Composite
+    IK_CompositeExtract,          // OpCompositeExtract
+    IK_ExtInst,                   // OpExtInst
+    IK_FunctionCall,              // OpFunctionCall
+    IK_GroupNonUniformBinaryOp,   // Group NonUniform binary operations
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/SpirvInstruction.h]: Also applies to GroupNonUniform enums

File: tools/clang/lib/SPIRV/SpirvBasicBlock.cpp [hunk 3]
```diff
@@ -17,7 +17,7 @@ SpirvBasicBlock::SpirvBasicBlock(uint32_t id, llvm::StringRef name)
       continueTarget(nullptr) {}
 
 bool SpirvBasicBlock::hasTerminator() const {
-  return !instructions.empty() && instructions.back()->isTerminator();
+  return !instructions.empty() && isa<SpirvTerminator>(instructions.back());
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvBasicBlock.cpp]: \o/

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h (hunk 1 above)
Comment (antiagainst) [reply tools/clang/include/clang/SPIRV/SpirvInstruction.h]: Good call. Done

File: tools/clang/include/clang/SPIRV/SpirvInstruction.h (hunk 2 above)
Comment (antiagainst) [reply tools/clang/include/clang/SPIRV/SpirvInstruction.h]: Done

**Pull request merged**